### ISSUE TITLE
I18n NumericField 

### ISSUE THREAD

**Issue by zigterman** [silverstripe/silverstripe-framework#2161]
When you use the NumericField and set your local to a country that uses a , instead of a . you get some problems.

The validator works, there is a check with Zend_Locale_Format and your local wil be used. But when you save the value a 0 is send to the database. The problem is in 
```php
public function dataValue() {
	return (is_numeric($this->value)) ? $this->value : 0;
}
```
the function expects a . instead of a ,. So the is_numeric returns false, you can use the check from Zend_Local_Format, but if your database expects a numer you need a . instead of a ,. I fixed this by making the function look like this:

```php
public function dataValue() {
	return Zend_Locale_Format::isNumber(trim($this->value), array('locale' => i18n::get_locale())) ? str_replace(',', '.', $this->value) : 0;
}
```

There still is one little problem now, when you reload your data in the field, it wil automatically use a . instead of ,. So when you save the data, you get een a validation error.

To fix this I override the setValue function to something like this:
```php
	public function setValue($value) {
		return parent::setValue(Zend_Locale_Format::toNumber($value));
	}
```

You have to do the first replace, when you load the data from the database there is a . instead of the expected ,.

I'm not working in php a lot, so maybe there is a better way to fix this. That's why I send a bug report and not a pull request.

**Comment by smindel** [comment]
This is an important issue for users of non-EN locales and I am concerned that we are about to cut corners again. I appreciate your taking on this matter and I want to put in my two cents:

Short:
- values stored in FormFields must not be localised but raw
- a dedicated setter for localised values must be added to the FormField API and this setter has to be used throughout the framework when setting values on FormFields to user supplied input
- alternatively the i18n class can be promoted to a full tier in the stack that form data has to pass through

Long:

Localization is supposed to help overcome differences in the representation of values for different locales in the view. While the user has every right to enter a VAT rate as "0,19" instead of "0.19", as developers we don't want to be bothered with all the babel in the business logic.

1. This being said, generally the API of a FormField has to provide dedicated setters and getters for raw data (which i prefer to be the primary representation of the data, because this is what we deal with in the business logic) as well as for the view (localise values for the rendered forms and *de*localise user provided localised input). This requires a bidirectonal lossless conversion between the representations. Basically the FormFields need the same mechanics as we have in the ORM, which translates back and forth between the raw data representation (php types) and the persistency representation (types of the storage engine)

2. We need certainty about what representation of the data we are dealing with. I strongly vote against a magic approach where the field is guessing, simply because it can not be done: '12.000' = (en/php)12 but 12000 in German. This would yield a conversion impasse in the magic inevitably which can be resolved by different setters/getters per representation or even representation aware getters/setters.

3. If we do this we might as well do it right cause we're not only talking about numbers here, but dates, times, and what not. While unlocalised numerics are a mere cosmetic nuisance most of the time, dates can be falsely interpreted much easier and their implementation is even more broken. This issue can not be fixed with a minor change but has an impact on each and every line in the core that deals with setting values to and getting values from fields in the entire framework.

4. It really matters to us and we're not a small minority (https://en.wikipedia.org/wiki/File:DecimalSeparator.svg) and don't be put of because you don't want to break things, it's broken already. BTW: there is another issue on the subject: #3354

Cheers

Andy

**Comment by dhensby** [comment]
@smindel I think generally there is probably an agreement with you when looked at objectively.

However, it is a bug in 3.1 and the fix in #3008 is one that will be able to be applied in 3.*.

Changing the setters and getters for value/presentation would have to go against v4 (master) and would take a long time for it to get released.

We have to take a compromise view at this point and (it appears that) #3008 will fix this problem in a way that will allow the bug to be patched for imminent versions of SS

**Comment by smindel** [comment]
I know very well what you are talking about. I need to give excuses to my customers everyday because localisation is broken to the point where workflows stop and data is lost. I don't want to be the party pooper but this issue deserves a sound solution not a patch. I sense that if we supply that patch we relieve the core team of the pressure to find the solution. I could think of a fix that can achieve both, set an example for the API and repair the flipping NumberField. Check out my approach above.
